states/kv/kv.go

```diff
@@ -0,0 +1,595 @@
+	}
+
+	// Check if the value being written needs to be empty plaeholder
+	byteValue, err := convertEmptyStringToByte(value)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to cast to byte (%s:%s) for putTiKVMeta", key, value))
+	}
+	err = txn.Set([]byte(key), byteValue)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to set value for key %s in putTiKVMeta", key))
+	}
+	return txn.Commit(ctx)
+}
+
+// Remove removes the input key.
+func (kv *txnTiKV) Remove(ctx context.Context, key string) error {
+	key = path.Join(kv.rootPath, key)
+
+	txn, err := kv.client.Begin()
+	if err != nil {
+		return errors.Wrap(err, "Failed to build transaction for removeTiKVMeta")
+	}
+
+	err = txn.Delete([]byte(key))
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("Failed to remove key %s in removeTiKVMeta", key))
+	}
+	return txn.Commit(ctx)
+}
+
+// RemoveWithPrefix removes the keys for the given prefix.
+func (kv *txnTiKV) RemoveWithPrefix(ctx context.Context, prefix string) error {
+	prefix = path.Join(kv.rootPath, prefix)
+	ctx, cancel := context.WithTimeout(ctx, RequestTimeout)
+	defer cancel()
+
+	startKey := []byte(prefix)
+	endKey := tikv.PrefixNextKey(startKey)
+	_, err := kv.client.DeleteRange(ctx, startKey, endKey, 1)
+	return err
+}
+
+func (kv *txnTiKV) removeWithPrevKV(ctx context.Context, key string) (*mvccpb.KeyValue, error) {
+	preV, err := kv.Load(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	err = kv.Remove(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	pkv := &mvccpb.KeyValue{
+		Key:   []byte(key),
+		Value: []byte(preV),
+	}
+	return pkv, nil
+}
+
+func (kv *txnTiKV) removeWithPrefixAndPrevKV(ctx context.Context, prefix string) ([]*mvccpb.KeyValue, error) {
+	keys, vals, err := kv.LoadWithPrefix(ctx, prefix)
+	if err != nil {
+		return nil, err
+	}
+	err = kv.RemoveWithPrefix(ctx, prefix)
+	if err != nil {
+		return nil, err
+	}
+	var kvs []*mvccpb.KeyValue
+	for i, key := range keys {
+		pkv := &mvccpb.KeyValue{
+			Key:   []byte(key),
+			Value: []byte(vals[i]),
+		}
+		kvs = append(kvs, pkv)
+	}
+	return kvs, nil
+}
+
+func (kv *txnTiKV) GetAllRootPath(ctx context.Context) ([]string, error) {
+	var apps []string
+	ss := kv.client.GetSnapshot(MaxSnapshotTS)
+	ss.SetScanBatchSize(SnapshotScanSize)
+
+	// Retrieve key-value pairs with the specified prefix
+	startKey := []byte(" ")
+	endKey := []byte("~")
+	for {
+		iter, err := ss.Iter(startKey, endKey)
+		if err != nil {
+			return nil, err
+		}
+		defer iter.Close()
+
+		if iter.Valid() {
+			key := string(iter.Key())
+			parts := strings.Split(key, "/")
+			if parts[0] != "" {
+				apps = append(apps, parts[0])
+			} else {
+				break
+			}
+			// next key, since '0' is the next ascii char of '/'
+			startKey = []byte(parts[0] + "0")
+		} else {
+			break
+		}
+	}
+	return apps, nil
+}
+
+func (kv *txnTiKV) BackupKV(base, prefix string, w *bufio.Writer, ignoreRevision bool, batchSize int64) error {
```

Comment:
NIT: maybe a better abstraction here is to implement an iterator for EtcdKV and tikv